src/v/cloud_roles/gcp_refresh_impl.cc

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_roles/gcp_refresh_impl.h"
+
+#include "bytes/iobuf_istreambuf.h"
+#include "cloud_roles/logger.h"
+#include "cloud_roles/request_response_helpers.h"
+#include "json/document.h"
+#include "json/istreamwrapper.h"
+
+#include <seastar/core/sleep.hh>
+
+static constexpr std::string_view token_path
+  = "/computeMetadata/v1/instance/service-accounts/default/token";
+
+/// This header must be sent to GCP metadata API along with each request
+struct metadata_flavor {
+    static constexpr std::string_view header_name = "Metadata-Flavor";
+
+    static constexpr std::string_view value = "Google";
+};
+
+struct gcp_response_schema {
+    static constexpr std::string_view expiry_field = "expires_in";
+    static constexpr std::string_view token_field = "access_token";
+};
+
+cloud_roles::gcp_refresh_impl::gcp_refresh_impl(
+  ss::sstring api_host,
+  uint16_t api_port,
+  s3::aws_region_name region,
+  ss::abort_source& as,
+  retry_params retry_params)
+  : refresh_credentials::impl(
+    std::move(api_host), api_port, std::move(region), as, retry_params) {}
+
+ss::future<cloud_roles::api_response>
+cloud_roles::gcp_refresh_impl::fetch_credentials() {
+    http::client::request_header oauth_req;
+
+    oauth_req.method(boost::beast::http::verb::get);
+    oauth_req.target(token_path.data());
+    oauth_req.set(
+      metadata_flavor::header_name.data(), metadata_flavor::value.data());
+
+    co_return co_await make_request(make_api_client(), std::move(oauth_req));
+}
+
+cloud_roles::api_response_parse_result
+cloud_roles::gcp_refresh_impl::parse_response(iobuf response) {
```

Comment:
so far I haven't see any tests (maybe they are at the end) but a lot of this code looks like it could have some unit tests.

Reply:
the unit tests for all these components are in a later commit https://github.com/redpanda-data/redpanda/pull/5015/commits/090dba4cf0f50cffd7cee224d0477d9a48702c5b 

I normally add unit tests along with or before writing the code but in this case I had to spend some time to learn the test libraries so I added them at the end after writing the code.

I plan to add more unit tests for some corner cases like intermittent network errors, but that needs enhancements to the http imposter to be able to run functions in response to request (right now we fix responses before the http server starts). I can do that after the http imposter has that functionality.